test/csuite/tiered_abort/main.c

```diff
@@ -564,6 +568,79 @@ handler(int sig)
     testutil_die(EINVAL, "Child process %" PRIu64 " abnormally exited", (uint64_t)pid);
 }
 
+/*
+ * verify_tiered --
+ *     Verify the expected locations of tiered objects.
+ */
+static void
+verify_tiered(WT_SESSION *session)
+{
+    struct stat sb;
+    WT_CONFIG_ITEM cval;
+    WT_CURSOR *metac;
+    uint32_t i, last, oldest;
+    int ret;
+    char buf[256], *key, *value;
+    const char *name;
+
+    testutil_check(session->open_cursor(session, "metadata:", NULL, NULL, &metac));
+    while ((ret = metac->next(metac)) != WT_NOTFOUND) {
+        testutil_check(ret);
+        testutil_check(metac->get_key(metac, &key));
+        if (WT_PREFIX_SKIP(key, "tiered:")) {
+            /*
+             * We have a top level entry for a tiered table. Get its configuration string and find
+             * the last id allocated (that should be in the local directory) and all others down to
+             * oldest should exist in the bucket directory. With minor exceptions, objects in the
+             * bucket directory should be removed from the local directory.
+             */
+            testutil_check(metac->get_value(metac, &value));
+            testutil_check(__wt_config_getones((WT_SESSION_IMPL *)session, value, "last", &cval));
+            last = (uint32_t)cval.val;
+            testutil_check(__wt_config_getones((WT_SESSION_IMPL *)session, value, "oldest", &cval));
+            oldest = (uint32_t)cval.val;
+            fprintf(
+              stderr, "VERIFY_TIERED: %s last %" PRIu32 " oldest %" PRIu32 "\n", key, last, oldest);
+            testutil_check(__wt_tiered_name_str(
+              (WT_SESSION_IMPL *)session, key, last, WT_TIERED_NAME_ONLY, &name));
+            /* Verify the latest object is in the local directory. */
+            testutil_check(__wt_snprintf(buf, sizeof(buf), "%s/%s", home, name));
+            ret = stat(buf, &sb);
+            testutil_assert(ret == 0);
+            /* Verify the latest object is not in the bucket directory. */
+            testutil_check(
+              __wt_snprintf(buf, sizeof(buf), "%s/%s/%s%s", home, BUCKET, BUCKET_PFX, name));
+            ret = stat(buf, &sb);
+            testutil_assert(ret != 0);
+            free((void *)name);
+            for (i = oldest; i < last; ++i) {
+                /* Verify earlier objects and their existence in the local directory. */
+                testutil_check(__wt_tiered_name_str(
+                  (WT_SESSION_IMPL *)session, key, i, WT_TIERED_NAME_ONLY, &name));
+                testutil_check(__wt_snprintf(buf, sizeof(buf), "%s/%s", home, name));
+                ret = stat(buf, &sb);
+                /*
+                 * Logged tables, i.e. "oplog" or "local" may be unable to remove the last object
+                 * from before the restart due to recovery applying log records. So if we get a stat
+                 * return that indicates the file exists, verify it is one of those tables.
+                 */
```

Comment:
Does this mean the files will eventually be removed, or is there a hole such that they will continue to exist in the local directory?